lib/checkunusedvar.cpp

```diff
@@ -1426,27 +1427,57 @@ void CheckUnusedVar::unusedStructMemberError(const Token *tok, const std::string
 bool CheckUnusedVar::isRecordTypeWithoutSideEffects(const Type* type)
 {
     // a type that has no side effects (no constructors and no members with constructors)
-    /** @todo false negative: check constructors for side effects */
-
     const std::pair<std::map<const Type *,bool>::iterator,bool> found=mIsRecordTypeWithoutSideEffectsMap.insert(
                 std::pair<const Type *,bool>(type,false)); //Initialize with side effects for possible recursions
     bool & withoutSideEffects=found.first->second;
     if (!found.second)
         return withoutSideEffects;
 
-    if (type && type->classScope && type->classScope->numConstructors == 0 &&
-        (type->classScope->varlist.empty() || type->needInitialization == Type::NeedInitialization::True)) {
-        for (std::vector<Type::BaseInfo>::const_iterator i = type->derivedFrom.begin(); i != type->derivedFrom.end(); ++i) {
-            if (!isRecordTypeWithoutSideEffects(i->type)) {
-                withoutSideEffects=false;
+    // unknown types are assumed to have side effects
+    if (!type || !type->classScope)
+    {
+        withoutSideEffects = false;
+        return withoutSideEffects;
+    }    
+
+    // non-empty constructors -> possible side effects
+    if (type->classScope->numConstructors != 0)
+    {
+        for (std::vector<Function>::const_iterator i = type->classScope->constructors.begin(); 
+            i != type->classScope->constructors.end(); i++)
+        {
+            if (!i->isBodyEmpty())
+            {
+                withoutSideEffects = false;
+                return withoutSideEffects;    
+            }
+        }
+    }
+
+    // check each variable type for side effects
+    if (!(type->classScope->varlist.empty() || type->needInitialization == Type::NeedInitialization::True))
+    {
+        for (std::list<Variable>::const_iterator i = type->classScope->varlist.begin(); 
+            i != type->classScope->varlist.end(); i++)
```

Comment:
use prefix increment to fix warnings in CI

Reply:
Already fixed in [this](https://github.com/danmar/cppcheck/pull/2692/commits/1995ed0ad0d790056335e1df0a71df6a093c149d) commit.
Seems like CI fail is not caused by my changes:
```
test/testnullpointer.cpp:2942:37: warning: Variable 'arg' is not assigned a value. [unassignedVariable]
            Library::ArgumentChecks arg;
                                    ^
test/testnullpointer.cpp:2955:37: warning: Variable 'arg' is not assigned a value. [unassignedVariable]
            Library::ArgumentChecks arg;
                                    ^
test/testsuppressions.cpp:513:22: warning: Unused variable: suppressions [unusedVariable]
        Suppressions suppressions;
```